src/exchanges/populate.ts

```diff
@@ -0,0 +1,331 @@
+import { pipe, tap, map } from 'wonka';
+import { Exchange, Operation } from '../types';
+import {
+  DocumentNode,
+  buildClientSchema,
+  visitWithTypeInfo,
+  TypeInfo,
+  FragmentDefinitionNode,
+  SelectionSetNode,
+  GraphQLSchema,
+  IntrospectionQuery,
+  FragmentSpreadNode,
+  GraphQLObjectType,
+  GraphQLInterfaceType,
+  ASTNode,
+} from 'graphql';
+import { visit } from 'graphql';
+
+interface ExchangeArgs {
+  schema: IntrospectionQuery;
+}
+
+/** An exchange for auto-populating mutations with a required response body. */
+export const populateExchange = ({
+  schema: ogSchema,
+}: ExchangeArgs): Exchange => ({ forward }) => {
+  const schema = buildClientSchema(ogSchema);
+  /** List of operation keys that have already been parsed. */
+  const parsedOperations = new Set<number>();
+  /** List of operation keys that have not been torn down. */
+  const activeOperations = new Set<number>();
+  /** Collection of fragments used by the user. */
+  let userFragments: UserFragmentMap = {};
+  /** Collection of type fragments. */
+  let typeFragments: TypeFragmentMap = {};
+
+  /** Handle mutation and inject selections + fragments. */
+  const handleIncomingMutation = (op: Operation) => {
+    if (op.operationName !== 'mutation') {
+      return op;
+    }
+
+    const activeSelections = Object.entries(typeFragments).reduce(
+      (state, [key, value]) => ({
+        ...state,
+        [key]: value.filter(s => activeOperations.has(s.key)),
+      }),
+      typeFragments
+    );
+
+    return {
+      ...op,
+      query: addFragmentsToQuery({
+        schema,
+        typeFragments: activeSelections,
+        userFragments: userFragments,
+        query: op.query,
+      }),
+    };
+  };
+
+  /** Handle query and extract fragments. */
+  const handleIncomingQuery = ({ key, operationName, query }: Operation) => {
+    activeOperations.add(key);
```

Comment:
Is this supposed to be below the escape hatch below? seems a little extreme to track all operations here if don't need all of them (in the case of unknown operations that may become an issue)

Mainly I'm concerned that the `activeOperations.delete(key)` for teardowns is in a different place, which makes it hard to track the above and the add/delete for teardowns

Reply:
Good catch - fixed